sdk/azcore/arm/resource_identifier.go

```diff
@@ -0,0 +1,315 @@
+//go:build go1.16
+// +build go1.16
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package arm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+)
+
+const (
+	providersKey             = "providers"
+	subscriptionsKey         = "subscriptions"
+	resourceGroupsLowerKey   = "resourcegroups"
+	locationsKey             = "locations"
+	builtInResourceNamespace = "Microsoft.Resources"
+)
+
+var (
+	RootResourceIdentifier = &ResourceIdentifier{
+		parent:       nil,
+		resourceType: TenantResourceType,
+		name:         "",
+	}
+
+	SubscriptionResourceType  = NewResourceType("Microsoft.Resources", "subscriptions")
+	ResourceGroupResourceType = NewResourceType("Microsoft.Resources", "resourceGroups")
+	TenantResourceType        = NewResourceType("Microsoft.Resources", "tenants")
+	ProvidersResourceType     = NewResourceType("Microsoft.Resources", "providers")
+)
+
+// ResourceIdentifier represents a resource ID
+type ResourceIdentifier struct {
+	parent            *ResourceIdentifier
+	subscriptionId    string
+	provider          string
+	resourceGroupName string
+	location          string
+	resourceType      ResourceType
+	name              string
+	isChild           bool
+
+	stringValue *string
+}
+
+func (id ResourceIdentifier) Parent() *ResourceIdentifier {
```

Comment:
I understand this was translated from C# but we should remove the getters and just export the necessary fields.

Reply:
I was thinking that we need to make this struct read-only and can only be initialized using the `ParseResourceIdentifier` function to make sure all the fields in this struct will be handled by our own logic

Reply:
This starts looking like C# IMO.

Reply:
I see your point, we can leave as-is for now.